### PR TITLE
[tests] add `torch.use_deterministic_algorithms` for XPU

### PR DESCRIPTION
**Pull request by faaany**
## What does this PR do?
There are tests that compare numerical difference of inference results between 2 separate model runs. These tests fail on XPU, because deterministic is not the default mode in oneDNN. Below is an example:

```python
import torch
import intel_extension_for_pytorch
from transformers import set_seed

set_seed(20)
model = torch.nn.Linear(768, 768, bias=True).to("xpu")

hs = torch.randn(768, 768).to("xpu")

query_states = model(hs)
query_states2 = model(hs)

re = (query_states == query_states2).all().item()
print(re)

# on XPU
>>>> False
# on CUDA&CPU
>>>> True
```
Although the difference in this example is from precision level of 1e-8, it will cause some some model tests fail. This PR fixes these failed tests by add `torch.use_deterministic_algorithms(True)`.

@ArthurZucker and @amyeroberts

### CODE REVIEW

**Comment by ydshieh** [comment]
Thanks! It's probably better to set this in a global way IMO, i.e. set it in a single place (only when we launch tests on a  XPU device)? WDYT?

**Comment by faaany** [comment]
Hi @ydshieh, that's correct! I almost forgot the `spec.py` file. I will put the function into `spec.py` and close this PR. Thanks for the review!

**Comment by faaany** [comment]
Hi @ydshieh , I just finished running all tests on XPU with `torch.use_deterministic_algorithms(True)` in `spec.py` and I am afraid that I have to add it in individual tests, because adding it in a global place will cause a lot of other tests fail.

**Comment by ydshieh** [comment]
@faaany 

Could you share what kinds of errors we have if we add `torch.use_deterministic_algorithms(True)` in `spec.py`?

**Comment by faaany** [comment]
> @faaany
> 
> Could you share what kinds of errors we have if we add `torch.use_deterministic_algorithms(True)` in `spec.py`?

sure, different kinds of errors, e.g. 

```bash
RuntimeError: linearIndex.numel() * sliceSize * nElemBefore == expandedValue.numel() INTERNAL ASSERT FAILED at "/build/intel-pytorch-extension/csrc/gpu/aten/operators/Indexing.cpp":1289, please report a bug to PyTorch. number of flattened indices did not match number of elements in the value tensor: 160 vs 5
```

```bash
RuntimeError: grid_sampler_2d_backward_xpu does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True)'. You can turn off determinism just for this operation, or you can use the 'warn_only=True' option, if that's acceptable for your application. You can also file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation.
```

```bash
RuntimeError: scatter_add_dpcpp_kernel does not have a deterministic implementation, but you set 'torch.use_deterministic_algorithms(True)'. You can turn off determinism just for this operation, or you can use the 'warn_only=True' option, if that's acceptable for your application. You can also file an issue at https://github.com/pytorch/pytorch/issues to help us prioritize adding deterministic support for this operation.
```

**Comment by faaany** [comment]
The majority of the error happens at "Variable._execution_engine.run_backward()".

**Comment by ydshieh** [comment]
Hi @faaany I would need a bit more context (in particular I don't have a XPU device to test) 🙏 

So I am wondering why adding `torch.use_deterministic_algorithms(True)` globally will make other (more) tests fail? Are such failures expected (i.e. they are **known** to not working with `torch.use_deterministic_algorithms(True)`?). 

From the description, IIRC, those extra tests (failed with `torch.use_deterministic_algorithms(True)`) would pass if `use_deterministic_algorithms` is not set to `True`?

**Comment by faaany** [comment]
Hi @ydshieh, I tried to add `torch.use_deterministic_algorithms(True)` in CUDA environment and those tests that are failed in XPU environment also fail in CUDA with the following message:

```bash
E       RuntimeError: Deterministic behavior was enabled with either `torch.use_deterministic_algorithms(True)` or `at::Context::setDeterministicAlgorithms(true)`, but this operation is not deterministic because it uses CuBLAS and you have CUDA >= 10.2. To enable deterministic behavior in this case, you must set an environment variable before running your PyTorch application: CUBLAS_WORKSPACE_CONFIG=:4096:8 or CUBLAS_WORKSPACE_CONFIG=:16:8. For more information, go to https://docs.nvidia.com/cuda/cublas/index.html#cublasApi_reproducibility
```
e.g. 
```bash
pytest tests/models/tapas/test_modeling_tapas.py::TapasModelTest::test_problem_types -rA
pytest tests/models/longformer/test_modeling_longformer.py::LongformerModelTest::test_for_question_answering
```

And my `spec.py` is like this:
```python
import torch
DEVICE_NAME = "cuda"
torch.use_deterministic_algorithms(True)
MANUAL_SEED_FN = torch.cuda.manual_seed
EMPTY_CACHE_FN = torch.cuda.empty_cache
DEVICE_COUNT_FN = torch.cuda.device_count
```

The only difference is that the error messages in CUDA are consistent saying that this operation is not deterministic and you should not set deterministic=True.

**Comment by ydshieh** [comment]
Hi @faaany 

Thanks for sharing. I am open to what is proposed in this PR, but I have one more question. If we do
```python
        if "xpu" in torch_device:
            torch.use_deterministic_algorithms(True)
```
in a test method, it actually changes  torch's property once after. All the subsequent tests run after that test will always have `deterministic` being `True` and should have the same issue as if we set in the first place.

We might not see that failures due to the luck (depending on the order of tests being run) but there is no guarantee.

In any case, since this is currently only XPU related, it's fine from my side if you want to move forward with this change.

However I would prefer to use a decorator like `require_deterministic_for_xpu` and it does

```
         if "xpu" in torch_device:
            torch.use_deterministic_algorithms(True)
```
and we just use decorator on some methods that requires it.

**Comment by faaany** [comment]
Hi @ydshieh , you are right! This function will have a global impact. If I put it in a test method, the following tests will be affected. 

Furthermore, I found that the function `torch.use_deterministic_algorithms()` have a flag called `warn_only`. So if I use `torch.use_deterministic_algorithms(mode=True, warn_only=True)` on CUDA, the failed tests I mentioned earlier will give a warning instead of failing. Unfortunately, this flag doesn't work well on XPU. I have reported the issue to the corresponding team and hope that it could be fixed.

However, back to our PR, I think `require_deterministic_for_xpu`  is still a better idea instead of applying this function to all tests due to its "large impact". So I updated my fix accordingly. Pls have a review.

**Comment by faaany** [comment]
There are some tests failed, but I think this should not be related to my fix.

**Comment by ydshieh** [comment]
Hi. Just to double check, do you intend to skip those tests for XPU? If so I am fine with the changes.

**Comment by ydshieh** [comment]
I mean, you can still try to set it to deterministic for XPU (in the decorator body) and use it (despite you might get some failures at some points as I mentioned). You might get lucky 😆 

But if skip is an option for you, OK for me too!

**Comment by faaany** [comment]
Yes, I think skip is a better option here, because there are only less than 10 out of 13508 tests affected by the deterministic algorithm. So treating them separately is better so other tests won't be affected. 😊

File: src/transformers/testing_utils.py [hunk 1]
```diff
@@ -917,6 +918,13 @@ def require_torch_bf16_cpu(test_case):
     )(test_case)
 
 
+def require_deterministic_for_xpu(test_case):
+    if is_torch_xpu_available():
+        return unittest.skipUnless(is_torch_deterministic(), "test requires torch to use deterministic algorithms")(
+            test_case
+        )
```
Comment (ydshieh) [inline src/transformers/testing_utils.py:925]: I believe that we will always need to return something in such functions.

Currently, if `is_torch_xpu_available` is not True, nothing is returned, and I think something will go wrong.

(got to say I didn't run myself, it's just what I see from other functions in this file)

File: src/transformers/utils/import_utils.py [hunk 2]
```diff
@@ -294,6 +294,18 @@ def is_torch_available():
...
+def is_torch_deterministic():
+    """
+    Check whether `torch.use_deterministic_algorithms()` is set to True or torch.set_deterministic_debug_mode() is set to 1 or 2
+    """
+    import torch
+
+    if torch.get_deterministic_debug_mode() == 0:
+        return False
+    else:
+        return True
```
Comment (ydshieh) [inline src/transformers/utils/import_utils.py:306]: not corresponds to the docstring: `torch.use_deterministic_algorithms` is not used in the code, but mentioned in the docstring.

**Comment by ydshieh** [comment]
Hi @faaany 

I still have 2 questions.

File: src/transformers/testing_utils.py (hunk 1 above)
Comment (faaany) [reply src/transformers/testing_utils.py:925]: what would be your suggestion for this case? I see 2 options:
- add an "else" clause and return unitest.skipUnless(False)(test_case)
- add a `pytest.marker`

File: src/transformers/utils/import_utils.py (hunk 2 above)
Comment (faaany) [reply src/transformers/utils/import_utils.py:306]: Should I change to "Check whether pytorch uses deterministic algorithms or torch.set_deterministic_debug_mode() is set to 1 or 2"?

File: src/transformers/testing_utils.py (hunk 1 above)
Comment (ydshieh) [reply src/transformers/testing_utils.py:925]: Add an `else` where just return `test_case` maybe?

File: src/transformers/utils/import_utils.py (hunk 2 above)
Comment (ydshieh) [reply src/transformers/utils/import_utils.py:306]: Oh `torch.use_deterministic_algorithms()` is a set function not get function.

> Check whether pytorch uses deterministic algorithms by looking if torch.set_deterministic_debug_mode() is set to 1 or 2"

how about this?

**Review by ydshieh — APPROVED** [review]
Thanks again @faaany . LGTM.

**Comment by faaany** [comment]
I think the failing tests are not caused by my fix. Could you retrigger the CI? Thx! @ArthurZucker

**Comment by ydshieh** [comment]
Try to rebase on main and push again 🤗

**Review by amyeroberts — APPROVED** [review]
@ydshieh @faaany Thanks both for iterating on this to find a solution, and thanks again @faaany for another improvement to our test suite!

**Pull request merged**